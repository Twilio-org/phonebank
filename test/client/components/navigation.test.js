import React from 'react';
import { mount, shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Navigation from '../../../public/src/components/navigation';

describe('<Navigation />', () => {
  const testUser = {
    first_name: 'Beyonce',
    last_name: 'Cat',
    email: 'mrcat@gmail.com',
    phone_number: '15555555555'
  };
  const linksLoggedIn = [
    { title: 'Account', href: `/account/1` },
    { title: 'Logout', href: '/logout' }
  ];
  const linksLoggedOut = [
    { title: 'Register', href: '/registration' },
    { title: 'Login', href: '/login' }
  ];
  const logout = jest.fn();
  const propsLoggedIn = {
    title: testUser.first_name,
    links: linksLoggedIn,
    userInfo: testUser,
    history: {},
    logout: logout
  };
  const propsLoggedOut ={
    title: 'Menu',
    links: linksLoggedOut
  };
  describe('rendering', () => {
    const wrapper = mount(
      <MemoryRouter>
        <Navigation {...propsLoggedOut}/>
      </MemoryRouter>
    );
    it('renders correctly', () => {
      const tree = renderer.create(
        <MemoryRouter>
          <Navigation {...propsLoggedOut} />
        </MemoryRouter>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('should render <Nav />', () => {
      expect(wrapper.find('Nav').length).toBe(1);
    });
    it('should render <NavDropdown />', () => {
      expect(wrapper.find('NavDropdown').length).toBe(1);
    });
    it('should render <Link />', () => {
      expect(wrapper.find('Link').length).toBeGreaterThan(1);
    });
  });

  describe('props', () => {
    const wrapper = shallow(<Navigation {...propsLoggedOut} />);

    it('should recieve "Menu" if user is logged out',() => {
      expect(wrapper.instance().props['title']).toBe('Menu');
    });
    it('should recieve "Login"/"Register" links if user is logged out',() => {
      expect(wrapper.instance().props['links']).toBe(linksLoggedOut);
    });
    it('should recieve name if user is logged in',() => {
      wrapper.setProps(propsLoggedIn);
      expect(wrapper.instance().props['title']).toBe('Beyonce');
    });
    it('should recieve "Login"/"Register" links if user is logged out',() => {
      wrapper.setProps(propsLoggedIn);
      expect(wrapper.instance().props['links']).toBe(linksLoggedIn);
    });
  });

  describe('onClick',() => {
    it('should call logoutOnClick() when "Logout" is clicked',() => {
      const wrapper = mount(
        <MemoryRouter>
          <Navigation {...propsLoggedIn}/>
        </MemoryRouter>
      );
      wrapper.find('a').last().find('a').simulate('click');
      expect(logout).toHaveBeenCalled();
    });
    it('should not render logout link if user logged out',() => {
      const wrapper = mount(
        <MemoryRouter>
          <Navigation {...propsLoggedOut}/>
        </MemoryRouter>
      );
      expect(wrapper.find('a').last().find('a').text()).not.toBe('Logout');
    });
  });

  describe('renderLinks()',() => {
    const wrapper = shallow(<Navigation {...propsLoggedOut} />);
    it('should return link items correctly',() => {
      const links = wrapper.instance().renderLinks();
      expect(Array.isArray(links)).toBe(true);
    });
  });
});
