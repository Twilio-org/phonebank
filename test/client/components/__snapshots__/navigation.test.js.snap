// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation /> rendering renders correctly 1`] = `
<ul
  className="nav pull-right"
  role={null}
>
  <li
    className="dropdown"
    style={undefined}
  >
    <a
      aria-expanded={false}
      aria-haspopup={true}
      className="dropdown-toggle"
      href="#"
      id="menu"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
    >
      Menu
       
      <span
        className="caret"
      />
    </a>
    <ul
      aria-labelledby="menu"
      className="dropdown-menu"
      role="menu"
    >
      <li
        active={undefined}
        onKeyDown={[Function]}
        onSelect={[Function]}
        role="presentation"
      >
        <a
          href="/registration"
          onClick={[Function]}
          role="menuitem"
          tabIndex="-1"
        >
          Register
        </a>
      </li>
      <li
        active={undefined}
        onKeyDown={[Function]}
        onSelect={[Function]}
        role="presentation"
      >
        <a
          href="/login"
          onClick={[Function]}
          role="menuitem"
          tabIndex="-1"
        >
          Login
        </a>
      </li>
    </ul>
  </li>
</ul>
`;
